[k8s] GPU Support and fractional resources for Kubernetes

romilbhardwaj opened the pull request:
Adds GPU support for Kubernetes clusters and allows Kubernetes SkyPilot clusters to be created with fractional resources (e.g., 2.5 CPUs).

How to use:
* **If on GKE** - Should work out of the box. Simply run `sky launch --gpus ...`
* **If on EKS or onprem** - Create an EKS cluster with `eksctl create cluster -f tests/kubernetes/eks_test_cluster.yaml`. To automatically label GPUs, run `python -m sky.utils.kubernetes.gpu_labeler`. Then run `sky launch --gpus ...`.

What's working:
- [x] Plumbing for requesting `nvidia.com/gpu` resources
- [x] GPU type selection for GKE
- [x] Convenience script to automatically label nodes with GPU types for EKS and onprem - `python -m sky.utils.kubernetes.gpu_labeler`
- [x] Toggle default image based on requested resources
- [x] Support for fractional resources

TODOs for later (will be filed as issues and deferred):
- [ ] GPU type selection for onprem GFD - blocked on finding a mapping of gpu name -> label.
- [ ] Use a lighter GPU image
- [ ] Docs for GPU support - will be included in #2324 

Tested (run the relevant ones):

- [x] Code formatting: `bash format.sh`
- [x] GPU tests: `pytest tests/test_smoke.py::test_job_queue`, `pytest tests/test_smoke.py::test_huggingface` on GKE GPU cluster
- [x] All Kubernetes smoke tests on EKS (incl. GPUs) `pytest tests/test_smoke.py --cloud kubernetes`

romilbhardwaj commented on examples/resnet_app.py:
```diff
@@ -4,7 +4,14 @@
-subprocess.run(f'cd {workdir} && git checkout 222cc86', shell=True, check=True)
+
+# Clone the repo locally to workdir
+subprocess.run(
+    'git clone https://github.com/concretevitamin/tpu '
+    f'{workdir} || true',
+    shell=True,
+    check=True)
+subprocess.run(f'cd {workdir} && git checkout 9459fee', shell=True, check=True)
```
From #2359

romilbhardwaj commented:
This is ready for review!

romilbhardwaj commented:
Update - we now have a script to automatically label GPU types on non-GKE clusters (e.g., onprem, EKS):

```
$ python -m sky.utils.kubernetes.gpu_labeler --help
usage: sky-gpu-labeler [-h] [--cleanup]

Labels GPU nodes in a Kubernetes cluster for use with SkyPilot. Operates by running a job on each node that parses nvidia-smi and patches the node with new labels. Labels created
are of the format skypilot.co/accelerators: <gpu_name>. Automatically creates a service account and cluster role binding with permissions to list nodes and create labels.

optional arguments:
  -h, --help  show this help message and exit
  --cleanup   delete all GPU labeler resources in the Kubernetes cluster.
```

Michaelvll commented on sky/clouds/kubernetes.py:
```diff
@@ -248,8 +267,10 @@ def get_accelerators_from_instance_type(
+        inst = KubernetesInstanceType.from_instance_type(instance_type)
+        return {
+            inst.accelerator_type: inst.accelerator_count
+        } if (inst.accelerator_count and inst.accelerator_type) else None
```
style nit: we need to explicitly check the `None` with `is not None` : )

Michaelvll commented on sky/clouds/kubernetes.py:
```diff
@@ -328,30 +393,35 @@ def _make(instance_list):
-            return False, 'Credentials not found - ' \
-                          f'check if {_CREDENTIAL_PATH} exists.'
+            return (False, 'Credentials not found - '
+                    'check if {_CREDENTIAL_PATH} exists.')
 
```
```suggestion
                    f'check if {_CREDENTIAL_PATH} exists.')

```

Michaelvll commented on sky/clouds/kubernetes.py:
```diff
@@ -368,10 +438,10 @@ def accelerator_in_region_or_zone(self,
                                       region: Optional[str] = None,
                                       zone: Optional[str] = None) -> bool:
-        # TODO(romilb): All accelerators are marked as not available for now.
+        # TODO(romilb): All accelerators are marked as available for now.
         #  In the future, we should return false for accelerators that we know
         #  are not supported by the cluster.
```
Should we do the similar check as the code starting at L331? We can probably refactor the checking out and call it here as well?

Michaelvll commented on tests/kubernetes/README.md:
````diff
@@ -13,30 +13,48 @@ To build this image locally and optionally push to the SkyPilot registry, run:
    * Note - only GKE standard clusters are supported. GKE autopilot clusters are not supported.
+   * Tip - to create an example GPU cluster for testing, this command will create a 2 node cluster with 1x T4 and another with 1x V100:
+     ```bash
+     gcloud beta container --project "skypilot-375900" clusters create "testcluster" --zone "us-central1-c" --no-enable-basic-auth --cluster-version "1.27.2-gke.1200" --release-channel "regular" --machine-type "n1-standard-8" --accelerator "type=nvidia-tesla-t4,count=1" --image-type "COS_CONTAINERD" --disk-type "pd-balanced" --disk-size "100" --metadata disable-legacy-endpoints=true --scopes "htt ...
````
Not sure if we want to show the project_id of our GCP project. Is it possible to remove it from the git histroy?

Michaelvll replied:
Also, it would be easier to make the project-id and cluster name configurable with env vars. Using `PROJECT_ID` and `CLUSTER_NAME`

Michaelvll replied:
Trying out this command with the following version but get the following error occurs:
```
Google Cloud SDK 443.0.0
beta 2023.08.11
core 2023.08.11
```

```
ERROR: (gcloud.beta.container.clusters.create) could not parse resource []
```

romilbhardwaj replied:
Great points!  Thanks, I'll clean up and remove our project_id.

The command seems to work for me on:
```
Google Cloud SDK 432.0.0
beta 2023.05.19
core 2023.05.19
gke-gcloud-auth-plugin 0.5.3
```
Let me try to upgrade my gcloud and try again.

romilbhardwaj replied:
Hmm, even after upgrading to the latest gcloud, the command seems to work for me.
```
Google Cloud SDK 443.0.0
beta 2023.08.11
core 2023.08.11
gke-gcloud-auth-plugin 0.5.5
```
Here are my [logs](https://gist.github.com/romilbhardwaj/9dcc71d871c044599f176dc24ce729aa). Are you sure you're pasting the command correctly?

romilbhardwaj replied:
I've updated this command to use envvars to detect project id and set cluster name :)

Michaelvll reviewed (commented):
Thank you for adding the GPU support @romilbhardwaj! This is very exciting! I went through the code except the ones in `gpu_labler` and the `test/kubernetes` (will check later). Left several comments, most of which are nits. 
I am now testing it on GKE. : )

Michaelvll commented on tests/kubernetes/README.md:
````diff
@@ -13,30 +13,48 @@ To build this image locally and optionally push to the SkyPilot registry, run:
+5. **If you want GPU support**, make sure you install GPU drivers by running:
+   ```bash
+   # If using COS based nodes (e.g., in the example above):
+   kubectl apply -f https://raw.githubusercontent.com/GoogleCloudPlatform/container-engine-accelerators/master/nvidia-driver-installer/cos/daemonset-preloaded.yaml
````
Is it possible to add this into some hint message when people try to run `sky launch --gpus V100` on the kubernetes cluster? Currently, it seems only showing timed out when it fails to get the resources on the kubnertest cluster.

romilbhardwaj replied:
I've updated `sky check` to return hints for:
1. If GPU drivers/nvidia device plugin is not installed (i.e., `nvidia.com/gpu` resource is not present on the cluster).
2. If GPU types are not labelled correctly (i.e., none of the labels in the `LABEL_FORMATTER_REGISTRY` are available on the cluster. E.g., `cloud.google.com/gke-accelerator` or `skypilot.co/accelerator` labels are not present.).

Example:
```
$ sky check
...
  Kubernetes: enabled
    Hint: Could not detect GPU resources (`nvidia.com/gpu`) in Kubernetes cluster. If this cluster contains GPUs, please ensure GPU drivers are installed on the node. Check if the GPUs are setup correctly by running `kubectl describe nodes` and looking for the nvidia.com/gpu resource. Please refer to the documentation on how to set up GPUs.
...
```

Would this be helpful?

Michaelvll commented on Dockerfile_k8s_gpu:
```diff
@@ -0,0 +1,56 @@
+
+# Set WORKDIR and initialize conda for sky user
+WORKDIR /home/sky
+RUN conda init
```
It seems in the GPU node created in the GKE. I am not able to run `nvidia-smi`.

```
nvida-msi
-bash: nvida-msi: command not found
```

Is it because some cuda driver issue?

romilbhardwaj replied:
Hmmm, are you running nvidia-smi inside a skypilot task (e.g., `sky exec myclus -- nvidia-smi`) or by ssh-ing into the cluster? 

If it is the latter (`ssh myclus` then `nvidia-smi`), then its a [known issue](https://github.com/skypilot-org/skypilot/issues/2287#issuecomment-1667221881) for now and @hemildesai is working on it :)

If it is the former, then it should work... can you share the output of `kubectl describe pods` and also run `kubectl exec -it <pod-name> /bin/bash` then `nvidia-smi` once you are inside the pod? (You can get the pod name by running `kubectl get pods`).

romilbhardwaj replied:
@Michaelvll - I've fixed and pushed the updated GPU image. It should work now if you try on a fresh GKE cluster :)

romilbhardwaj commented on sky/clouds/kubernetes.py:
```diff
@@ -123,6 +136,14 @@ def from_resources(cls,
     def __str__(self):
         return self.name
 
+    @classmethod
+    def _format_count(cls, num: Union[float, int]) -> str:
+        """Formats a float to not show decimal point if it is a whole number"""
+        if isinstance(num, int):
+            return str(num)
+        return '{:.0f}'.format(num) if num.is_integer() else '{:.1f}'.format(
+            num)
```
That's a good idea - I've moved this to `common_utils` so future usage can directly invoke `common_utils.format_float`

romilbhardwaj commented on sky/clouds/kubernetes.py:
```diff
@@ -293,19 +314,63 @@ def make_deploy_resources_variables(
+                        'node labels of either of these formats: '
+                        f'{supported_formats}. Please refer to '
+                        'the documentation on how to set up node labels.'
+                        f'{suffix}')
```
That's a great suggestion, thanks! I have refactored GPU detection + label selection to be a separate method (`kubernetes_utils.get_gpu_label_key_value`), which is now also invoked earlier in `_get_feasible_launchable_resources`. This allows the optimizer to not include Kubernetes if the GPU is not present on the cluster.

romilbhardwaj commented on sky/backends/backend_utils.py:
```diff
@@ -885,15 +885,18 @@ def write_cluster_config(
+    #  * The region/zones requested does not appear in the catalog. It can be
+    #    triggered if the user changed the catalog file while there is a cluster
+    #    in the removed region/zone.
+    #  * GPUs are requested in a Kubernetes cluster but the cluster does not
+    #    have nodes labeled with GPU types.
```
That's a good point. With the refactor you suggested, we now check GPU availability earlier in `_get_feasible_launchable_resources`. Now, under normal operation, our Kubernetes cloud will never throw `ResourceUnavailableError` in `make_deploy_resources_variables`.

 However, if something goes really wrong (something I don't foresee currently), it's still _possible_ it will raise `ResourceUnavailableError` from the underlying `get_gpu_label_key_value` method. Does that work?

romilbhardwaj commented on sky/clouds/kubernetes.py:
```diff
@@ -167,8 +188,10 @@ class Kubernetes(clouds.Cloud):
 
-    IMAGE = 'us-central1-docker.pkg.dev/' \
-            'skypilot-375900/skypilotk8s/skypilot:latest'
+    IMAGE_CPU = ('us-central1-docker.pkg.dev/'
+                 'skypilot-375900/skypilotk8s/skypilot:latest')
+    IMAGE_GPU = ('us-central1-docker.pkg.dev/skypilot-375900/'
+                 'skypilotk8s/skypilot-gpu:latest')
```
Ah that's a good idea... since this would require adding a service catalog skeleton for Kubernetes, I'm thinking of doing it in a separate PR. I have filed #2430 to track this

romilbhardwaj commented:
@Michaelvll - I've implemented the suggested refactor and it looks much cleaner now - thanks for the great feedback! This should be ready for another look.

- [x] `pytest tests/test_smoke.py --kubernetes`

Michaelvll commented on sky/clouds/kubernetes.py:
```diff
@@ -328,30 +367,46 @@ def _make(instance_list):
 
         # Currently, handle a filter on accelerators only.
         accelerators = resources.accelerators
+        default_instance_type = Kubernetes.get_default_instance_type(
```
Should we increase the `DEFAULT_MEMORY_CPU_RATIO` when the GPU is involved? Users using SkyPilot to launch GPU tasks are likely running training or serving workloads that may require larger memory size than the same size as the number of cores of CPUs. Can we set it to 4 instead? Similar to the other clouds: https://github.com/skypilot-org/skypilot/blob/d00685e6c978435a656386fa05ff5bf0af1bcfb4/sky/clouds/service_catalog/gcp_catalog.py#L60

Michaelvll commented on sky/backends/cloud_vm_ray_backend.py:
```diff
@@ -2865,12 +2866,37 @@ def _sync_file_mounts(
+        See https://github.com/skypilot-org/skypilot/issues/2287 for
+        more details.
+        """
+        if isinstance(handle.launched_resources.cloud, clouds.Kubernetes):
+            temp_envs = copy.deepcopy(task.envs)
+            cloud_env_vars = handle.launched_resources.cloud.query_env_vars(
+                handle.cluster_name)
+            task.update_envs(cloud_env_vars)
+
+            # Re update the envs with the original envs to give priority to
+            # the original envs.
+            task.update_envs(temp_envs)
```
I would prefer to include those necessary env vars in the `~/.bashrc` instead of the users' task env vars. 
Reason: These env vars seem to be global for all the tasks submitted to the same VM, and for the user who would like to `ssh` into the VM, the `nvidia-smi` will not work if we only enable the env vars for the tasks.

romilbhardwaj replied:
That's a good point! Accessing SSH vars is being worked on by @hemildesai in https://github.com/landscapepainter/skypilot/pull/1, and we plan to ship it in #2435.  If you think this is urgent, I can to add it `~/.bashrc` here as a stopgap fix and remove it in #2435.

------
More context - 

The challenge we are trying to solve here is that environment variables from Kubernetes are dynamic and change over time. 

One example, as #2287 mentions, is service discovery in Kubernetes. Kubernetes populates service IPs and ports as environment variables. These IPs and ports may change or new ones may be added, which requires us to query environment variables dynamically every time.

For SSH too, we want to query environment variables dynamically and populate the environment. In his PR, @hemildesai is doing so by adding a ProxyCommand which query envvars whenever a SSH connection is initiated and updates the environment variables in `/etc/environment`.

Michaelvll replied:
Thanks for the detailed explanation!

One question I have is that, if the env vars are dynamic and the user ssh into the VM, and try to access the env var after a while, does that mean the env vars can be staled?

I think it should be fine to let the k8s related variables to be added in the future. One thing I think that is important is to make sure the `nvidia-smi` works through ssh, as it might be very confusing that `ssh cluster-name` and `nvidia-smi` fail to work, even though the VM is launched with GPU resources.

I am guessing this may involve some nvidia-smi related PATH to be exported?

romilbhardwaj replied:
Totally agree - we should make `nvidia-smi` work for `ssh cluster-name`. This is fixed by #2435, and we plan to merge that PR also soon before we start advertising Kubernetes support, so I'll leave it out from this PR :) I have filed #2453 to make sure it gets fixed. Would that be ok?

> One question I have is that, if the env vars are dynamic and the user ssh into the VM, and try to access the env var after a while, does that mean the env vars can be staled?

Yes (though very rare), some envvars can be stale (e.g., if the service port gets changed by the user). These semantics work for Kubernetes applications because the pod will be automatically restarted if it crashes due to bad envvars, and the restarted pods will fetch latest envvars.

Michaelvll replied:
Sounds good! Thanks for the explanation : )

Michaelvll reviewed (approved):
Thanks for the quick update @romilbhardwaj! I just tested with a manually created GKE cluster with V100 and T4 in the node pool, and this PR magically select the correct node and create the cluster for me. I am now testing the commands in the README. : )

Please correct me if I am wrong, but one thing we might want to include in our document is that the node configuration for CPUs in the node pool may need to be a bit larger than the number of CPUs a user would normally need, as the kubernetes cluster may reserve some CPU resources for the control plane.

Michaelvll commented on Dockerfile_k8s_gpu:
```diff
@@ -0,0 +1,58 @@
+# TODO(romilb) - The base image used here (ray) is very large (11.4GB).
+#  as a result, this built image is about 13.5GB. We need to pick a lighter base
+#  image.
+FROM rayproject/ray:2.4.0-gpu
```
nit: should we explicitly use `ray:2.4.0-py310-gpu` as we are using py310 for other clouds. : )

romilbhardwaj replied:
Done! I've updated both the CPU and GPU images.

romilbhardwaj commented on sky/templates/kubernetes-ray.yml.j2:
```diff
@@ -181,16 +187,9 @@ available_node_types:
-              # the object store size is not set manually, ray will
-              # allocate a very large object store in each pod that may
-              # cause problems for other pods.
-              cpu: {{cpus}}
-              memory: {{memory}}G
```
[This article](https://sysdig.com/blog/kubernetes-limits-requests/) explains limits vs requests nicely, but in our context, there are two key reasons for removing limits:
1) For CPU, setting limits will block tasks for using idle CPU capacity. For example, if a task requests 2 CPUs and the node has 6 more idle CPUs, the task will be able to use only 2 CPUs worth of clock cycles even if it needs more.
2) For memory, if a pod exceeds memory limits, Kubernetes will immediately OOM kill it, even if there is unused memory. On the other hand, not setting memory limits does not perform OOM killing until there there are other pods whose `requests` have not been met.

romilbhardwaj commented on sky/clouds/kubernetes.py:
```diff
@@ -328,30 +367,46 @@ def _make(instance_list):
+        # Check if requested accelerator type is supported in the cluster.
+        # If not, this will raise ResourcesUnavailableError.
+        try:
+            _, _ = kubernetes_utils.get_gpu_label_key_value(acc_type)
```
Done! Added a check for CPU and memory too. 

Additionally, if `SKYPILOT_DEBUG=1`, we also show a debug message to help figure out max resources (CPU/Memory) and GPU types available on the cluster.

Examples:
```
(base) ➜  ~ sky launch -c test2 --gpus X1 --cloud kubernetes
D 08-23 19:07:16 optimizer.py:250] #### Task(run=<empty>)
D 08-23 19:07:16 optimizer.py:250]   resources: Kubernetes({'X1': 1}) ####
D 08-23 19:07:18 kubernetes.py:414] Instance type 2CPU--8GB--1X1 does not fit in the Kubernetes cluster. Reason: Could not find any node in the Kubernetes cluster with X1 GPU. Please ensure at least one node in the cluster has X1 GPU and node labels are setup correctly. Please refer to the documentation for more.  Available GPUs on the cluster: {'nvidia-tesla-t4'}

...

sky.exceptions.ResourcesUnavailableError: No launchable resource found for task Task(run=<empty>)
  resources: Kubernetes({'X1': 1}).
This means the kubernetes cluster does not contain any resources that satisfy this request.
To fix: relax or change the resource requirements.
Hint: 'sky show-gpus --all' to list available accelerators.
      'sky check' to check the enabled clouds.
```

Another example:
```
(base) ➜  ~ sky launch -c test2 --cpus 1000+
D 08-23 19:10:43 optimizer.py:250] #### Task(run=<empty>)
D 08-23 19:10:43 optimizer.py:250]   resources: <Cloud>(cpus=1000+) ####
D 08-23 19:10:45 kubernetes.py:383] Instance type 1000CPU--1000GB does not fit in the Kubernetes cluster. Reason: No nodes found with enough CPU and/or memory. Maximum resources found on a single node: 8.0 CPUs, 29.4G Memory

...

sky.exceptions.ResourcesUnavailableError: No launchable resource found for task Task(run=<empty>)
  resources: <Cloud>(cpus=1000+).
This means the catalog and kubernetes cluster does not contain any resources that satisfy this request.
To fix: relax or change the resource requirements.
Hint: 'sky show-gpus --all' to list available accelerators.
      'sky check' to check the enabled clouds.
```

romilbhardwaj commented:
Thanks @Michaelvll! Your comments were super helpful in improving the UX. I've addressed most of them  and also added early filtering based on CPU/Mem in addition to GPUs (as you [mentioned](https://github.com/skypilot-org/skypilot/pull/2328#discussion_r1300507046)).

Running smoke tests:
- [ ] `pytest test_smoke.py`
- [ ] `pytest test_smoke.py --kubernetes`

Michaelvll commented on sky/utils/kubernetes_utils.py:
```diff
@@ -147,6 +159,89 @@ def get_kubernetes_nodes() -> List[Any]:
+            if node_cpus > max_cpu:
+                max_cpu = node_cpus
+                max_mem = node_memory_gb
+            if node_cpus >= candidate_instance_type.cpus and \
+                    node_memory_gb >= candidate_instance_type.memory:
```
nit
```suggestion
            if (node_cpus >= candidate_instance_type.cpus and
                    node_memory_gb >= candidate_instance_type.memory):
```

Michaelvll commented on sky/utils/kubernetes_utils.py:
```diff
@@ -147,6 +159,89 @@ def get_kubernetes_nodes() -> List[Any]:
+    """
+
+    def check_cpu_mem_fits(
+            candidate_instance_type: kubernetes_cloud.KubernetesInstanceType,
```
nit: Can we move the `KubernetesInstanceType` to this utils file, so we can import this module in the `clouds.Kubernetes` instead of the other way around?

Not feeling strongly about this, just that it feels a bit surprising to me to see the cloud being imported in the utils module.

romilbhardwaj replied:
Sounds good!

Michaelvll commented on sky/utils/kubernetes_utils.py:
```diff
@@ -147,6 +159,89 @@ def get_kubernetes_nodes() -> List[Any]:
+        ]
+        assert len(gpu_nodes) > 0, 'GPU nodes not found'
+        # Check if the CPU and memory requirements are met on at least one node
+        fits, reason = check_cpu_mem_fits(k8s_instance_type, gpu_nodes)
+        if not fits:
+            if reason is not None:
+                reason_prefix = (f'GPU nodes with {acc_type} do not have '
+                                 'enough CPU and/or memory. ')
+                reason = reason_prefix + reason
+            return fits, reason
+        else:
+            return fits, reason
+    else:
+        # If no GPU is requested, check if CPU and memory requirements are met
+        # on at least one node.
+        fits, reason = check_cpu_mem_fits(k8s_instance_type, nodes)
+        if not fits:
+            if reason is not None:
+                reason_prefix = 'No nodes found with enough CPU and/or memory. '
+                reason = reason_prefix + reason
+            return fits, reason
+        else:
+            return fits, reason
```
nit: Seems most of this part are similar, should we combine them?

romilbhardwaj replied:
Ah good point - I've cleaned this up

Michaelvll reviewed (commented):
Thanks for the quick update @romilbhardwaj! It looks awesome to me. Left several comments.

Michaelvll commented on sky/clouds/kubernetes.py:
```diff
@@ -366,33 +366,59 @@ def _make(instance_list):
+
+        gpu_task_cpus = chosen_inst.cpus
+        # Special handling to bump up memory multiplier for GPU instances
+        gpu_task_memory = float(resources.memory) \
```
Oops, just saw the fix in 1741581, seems working now.

Michaelvll commented on sky/utils/kubernetes_utils.py:
```diff
@@ -0,0 +1,468 @@
+    return nodes
+
+
+def check_instance_fits(instance: str) -> Tuple[bool, Optional[str]]:
```
Just a thought for discussion: maybe in the future, we can have a `kubernetes_catalog` module, where we use the similar APIs as the other cloud, but calling these online checking methods to check the availability. : )

romilbhardwaj replied:
Ahh good point.. we started with having a hardcoded CSV catalog, then converged onto this design. Perhaps your suggestion of having a `kubernetes_catalog` is a good middle ground. I've left a TODO under `get_default_instance_type` for this :)

romilbhardwaj commented on sky/clouds/kubernetes.py:
```diff
@@ -366,33 +366,59 @@ def _make(instance_list):
-            # If GPU not found, return empty list.
-            return [], []
-        default_inst = KubernetesInstanceType.from_instance_type(
+        # Get default instance type to populate CPU/memory.
+        default_instance_type = Kubernetes.get_default_instance_type(
+            cpus=resources.cpus,
+            memory=resources.memory,
+            disk_tier=resources.disk_tier)
+
```
Ah that's a good point - the logic is a bit messy here. I have cleaned it up now.

Michaelvll commented on tests/test_optimizer_dryruns.py:
```diff
@@ -677,15 +679,25 @@ def _test_optimize_speed(resources: sky.Resources):
 
 def test_optimize_speed(enable_all_clouds, monkeypatch):
-    _test_optimize_speed(sky.Resources(cpus=4))
+    _test_optimize_speed(_make_resources(monkeypatch, cpus='4+'))
```
We are using `enable_all_clouds` in this test, which should have similar functionality as `_make_resources`. Should we either remove that fixture from the argument, or add the kubernetes related patch to `enable_all_clouds` as well?

romilbhardwaj replied:
Ah good catch, thanks! I've added the patching to `enable_all_clouds` and reverted the changes here.

Michaelvll reviewed (approved):
LGTM! Thanks for the fix @romilbhardwaj! Excited to see the GPU support being added.
